common/init_llvm.cpp

```diff
@@ -8,16 +8,27 @@
 
 namespace Carbon {
 
-InitLLVM::InitLLVM(int& argc, char**& argv) : init_llvm(argc, argv) {
+// Defined in all_llvm_targets.cpp. Will be null if that file is not linked in.
+[[gnu::weak]] extern auto InitLLVMTargets() -> void;
+
+InitLLVM::InitLLVM(int& argc, char**& argv)
+    : init_llvm(argc, argv), args(argv, argv + argc) {
+  // LLVM assumes that argc and argv won't change, and registers them with an
+  // `llvm::PrettyStackTraceProgram` that will crash if an argv element gets
+  // nulled out, which for example `testing::InitGoogleTest` does. So make a
+  // copy of argv for use by the program to satisfy LLVM's assumptions.
+  argc = args.size();
+  argv = args.data();
```

Comment:
Should this actually be handing `init_llvm` its own values, rather than passing a reference?

I'm asking because these changes to `argc` and `argv` looked odd to me: it's changing the values after they're passed to `init_llvm`. That seems inconsistent with the comments above.

Won't this undo the Windows-specific changes that `init_llvm` does? If not, does absl do the utf8 translation? Does it result in googletest still being able to modify the argc/argv seen by stack tracing?

Reply:
> Should this actually be handing `init_llvm` its own values, rather than passing a reference?

We could do that, but it's awkward because we'd need to create some extra state to hold them before calling the constructor of the field. I don't think it's worth it.

> Won't this undo the Windows-specific changes that `init_llvm` does?

No: `args` is copied out of `argv, argv+argc` after the `init_llvm` constructor "fixes" the `argc` and `argv` values, so this is preserving the things that `llvm::InitLLVM` did, not undoing them.

> If not, does absl do the utf8 translation, potentially leading to both trying to do a conflicting transform?

No, it can't: `argc` and `argv` are already broken before you reach `main` on Windows because they get converted from Unicode into the ANSI code page, not into UTF-8. There's no way to fix them by re-encoding, because a lossy translation was already performed. `llvm::InitLLVM` throws away `argc` and `argv` on Windows and recomputes them based on the original Unicode command line.

The absl API is required to process the `argc` and `argv` you give it, and not go off and work out its own ones (because the things you pass in might not be the original command line), so if we want to give it UTF-8 arguments, we need to do that before calling into absl's flag parser, which LLVM is now doing for us.

> Does it result in googletest still being able to modify the argc/argv seen by stack tracing?

No, which is a bug fix: we want the pretty stack trace entry to show the original command line arguments given to the program, not the result of them being modified as gtest eats the arguments that it knows about.

Reply:
Thanks for the explanation.

Do you want suggestions for changes to the comments, or will you take another look at those?

> No, which is a bug fix: we want the pretty stack trace entry to show the original command line arguments given to the program, not the result of them being modified as gtest eats the arguments that it knows about.

The argv now points at the same address that's later passed on to googletest, doesn't it?

Reply:
Comments updated.